src/uucore/src/lib/mods/backup_control.rs

```diff
@@ -37,22 +33,164 @@ pub fn determine_backup_suffix(supplied_suffix: Option<&str>) -> String {
+///         show_usage_error!("{}", err);
+///         return 1;
+///     },
+///     Ok(mode) => mode,
+/// };
+/// ```
+pub fn determine_backup_mode(
+    sopt_present: bool,
+    lopt_present: bool,
+    lopt_value: Option<&str>,
+) -> Result<BackupMode, String> {
+    if lopt_present {
+        // Use method to determine the type of backups to make. When this option
+        // is used but method is not specified, then the value of the
+        // VERSION_CONTROL environment variable is used. And if VERSION_CONTROL
+        // is not set, the default backup type is ‘existing’.
+        if let Some(method) = lopt_value {
+            // Second argument is for the error string that is returned.
+            match_method(method, "backup type")
+        } else if let Ok(method) = env::var("VERSION_CONTROL") {
+            // Second argument is for the error string that is returned.
+            match_method(&method, "$VERSION_CONTROL")
+        } else {
+            Ok(BackupMode::ExistingBackup)
+        }
+    } else if sopt_present {
+        // the short form of this option, -b does not accept any argument.
+        // Using -b is equivalent to using --backup=existing.
+        Ok(BackupMode::ExistingBackup)
+    } else {
+        // No option was present at all
+        Ok(BackupMode::NoBackup)
+    }
+}
+
+/// Match a backup option string to a `BackupMode`.
+///
+/// The GNU manual specifies that abbreviations to options are valid as long as
+/// they aren't ambiguous. This function matches the given `method` argument
+/// against all valid backup options (via `starts_with`), and returns a valid
+/// [`BackupMode`] if exactly one backup option matches the `method` given.
+///
+/// `origin` is required in order to format the generated error message
+/// properly, when an error occurs.
+///
+///
+/// # Errors
+///
+/// If `method` is ambiguous (i.e. may resolve to multiple backup modes) or
+/// invalid, an error is returned. The error contains the formatted error string
+/// which may then be passed to the [`show_usage_error`] macro.
+fn match_method(method: &str, origin: &str) -> Result<BackupMode, String> {
+    let x = vec![
+        "simple", "never", "numbered", "t", "existing", "nil", "none", "off",
+    ];
+
+    let matches: Vec<&&str> = x.iter().filter(|val| val.starts_with(method)).collect();
+    if matches.len() == 1 {
+        match *matches[0] {
+            "simple" | "never" => Ok(BackupMode::SimpleBackup),
+            "numbered" | "t" => Ok(BackupMode::NumberedBackup),
+            "existing" | "nil" => Ok(BackupMode::ExistingBackup),
+            "none" | "off" => Ok(BackupMode::NoBackup),
+            _ => panic!(), // cannot happen as we must have exactly one match
+                           // from the list above.
         }
     } else {
-        BackupMode::NoBackup
+        let error_type = if matches.len() == 0 {
+            "invalid"
+        } else {
+            "ambiguous"
+        };
+        Err(format!(
```

Comment:
Also I'm not quite happy with the formatting of this bit in the code, as it ruins the indentation. I'd be thankful for suggestions on how to improve it!

Reply:
I think it's fine. The only thing I could think of is to move the "Valid arguments..." part to a constant, but I don't think that's necessary or useful.